File: apps/backend/services/nfts.go
package services

import (
	"encoding/json"
	"fmt"
	"net/http"
	"neura-nft/config"
	"neura-nft/constants"
	"neura-nft/models"
)

func GetExploreNFTs(address, category string) ([]models.AlchemyNFT, []models.AlchemyNFT, error) {
	baseUrl := fmt.Sprintf("https://eth-mainnet.g.alchemy.com/nft/v2/%s", config.AlchemyApiKey)

	// ---- Owned NFTs
	ownedUrl := fmt.Sprintf("%s/getNFTs?owner=%s", baseUrl, address)

	ownedResp, err := http.Get(ownedUrl)
	if err != nil {
		return nil, nil, err
	}
	defer ownedResp.Body.Close()

	var ownedData models.AlchemyResponse
	if err := json.NewDecoder(ownedResp.Body).Decode(&ownedData); err != nil {

		return nil, nil, err
	}

	// ---- Trending NFTs
	contracts := constants.NftCategories[category]

	var trending []models.AlchemyNFT
	for _, c := range contracts {
		url := fmt.Sprintf("%s/getNFTsForCollection?contractAddress=%s&withMetadata=true", baseUrl, c)

		resp, err := http.Get(url)
		if err != nil {
			fmt.Println("❌ Collection fetch error:", err)
			continue
		}

		var colData models.AlchemyResponse
		if err := json.NewDecoder(resp.Body).Decode(&colData); err != nil {
		} else {
			for _, nft := range colData.Nfts {
				if len(nft.Media) > 0 && (nft.Media[0].Gateway != "" || nft.Media[0].Thumbnail != "") {
					trending = append(trending, nft)
				}
			}
		}
		resp.Body.Close()
	}

	return ownedData.OwnedNfts, trending, nil
}

File: apps/backend/src/services/nfts.ts
import { NftCategories } from "@/constants/nfts";
import { endpoints } from "@/constants/urls";
import axios from "axios";

export async function getExploreNFTs(address: string, category: string) {
  // --- Owned NFTs
  const { data: ownedData } = await axios.get(
    endpoints.alchemy.getOwnedNfts(address),
  );
  const owned = ownedData.ownedNfts || [];

  // --- Trending NFTs
  const contracts = NftCategories[category] || [];
  const trending = [];

  for (const contract of contracts) {
    try {
      const { data: colData } = await axios.get(
        endpoints.alchemy.getCollectionNfts(contract),
      );

      if (colData?.nfts) {
        for (const nft of colData.nfts) {
          const mediaUrl = nft.media?.[0]?.gateway || nft.media?.[0]?.thumbnail;

          if (mediaUrl && !mediaUrl.includes("ipfs")) {
            trending.push(nft);
          }
        }
      }
    } catch (err) {
      console.error("❌ Collection fetch error:", err);
    }
  }

  return { owned, trending };
}
